### ISSUE TITLE
Nodes must not produce blocks when they are significantly behind

### ISSUE THREAD

**Issue by nfrisby** [input-output-hk/ouroboros-network#1631]
[I'll amend this description later.]

From @edsko [on Slack](https://input-output-rnd.slack.com/archives/GCW2S19CM/p1581361972271700):

> [T]he node kernel must not produce any blocks unless the anchor of its fragment (not its tip) is within 2k slots from the wallclock. In other words, don't produce a block if your oldest immutable block is more than 2k slots old. There is a few ways to see this:
> 1. Don't produce any blocks when you are way behind
> 2. Attempt to enforce the k-blocks-in-2k-slots
> 3. Don't become your own adversary
> 
> Point (3) is related to [Ouroboros Genesis]: when we are behind, we don't want to adopt chains at all, but instead collect chains and pick the highest density (we don't do this yet of course). But this also implies we shouldn't produce our own blocks, because if we did, we'd become our own adversary.

I'm working to implement this now, as part of PR #1506 (since only that PR has so far exercised the problem).

At this point, I believe the rule is actually: the node must not forge a block such that its new chain fragment's anchor (ie `lastSlot` of its current chain fragment) would be more than 2k slots behind that block. There's also a off-by-1 issue here if the anchor is an EBB -- I haven't yet sorted that out.

-----

Background: many parts of a node assume there are at least k blocks in every 2k slot span. One place that invariant can be directly violated is if the `NodeKernel` forges a new block before it has received the most recent blocks. We have seen a test case where doing so introduces a wedge: once a Byron node adopts a block whose slot is more than 2k ahead of the beginning of the rest of the net's actual fork, it will never be able to switch to the fork; `MsgRollForward` and eventually `MsgFindIntersect` will always fail.

Until now, we have been relying on "the protocol" to ensure the invariant, but this Issue identifies that at least the `NodeKernel` needs to "do its part". We might want to revisit some of the APIs in light of this: e.g. the lower limit on the anachronistic ledger view is what `NodeKernel` should actually respect, and right now the types hide the fact that it can always reach up to 2k slots backwards.

**Comment by edsko** [comment]
Classifying this as medium priority since in reality this is not likely to cause issues.

**Comment by nfrisby** [comment]
Here are my notes about the repro we investigated that motivated this.

```
      prop_simple_real_pbft_convergence ProduceEBBs (SecurityParam 3) TestConfig
        { numCoreNodes = NumCoreNodes 5
        , numSlots     = NumSlots 40
        , nodeJoinPlan = NodeJoinPlan $ Map.fromList
          [ (CoreNodeId 0, SlotNo 0)
          , (CoreNodeId 1, SlotNo 0)
          , (CoreNodeId 2, SlotNo 0)
          , (CoreNodeId 3, SlotNo 33)
          , (CoreNodeId 4, SlotNo 33)
          ]
        , nodeRestarts = noRestarts
        , nodeTopology = NodeTopology $ Map.fromList   -- linear
          [ (CoreNodeId 0, Set.fromList [])
          , (CoreNodeId 1, Set.fromList [CoreNodeId 0])
          , (CoreNodeId 2, Set.fromList [CoreNodeId 1])
          , (CoreNodeId 3, Set.fromList [CoreNodeId 2])
          , (CoreNodeId 4, Set.fromList [CoreNodeId 3])
          ]
        , slotLengths  = singletonSlotLengths (SlotLength 1)
        , initSeed     = Seed (8633299106886139239,11146688890796126320,14792158977094571234,8638362867401803123,17464005238248054424)
```

When c3 and c4 join in slot 33, the existing net all has selected the chain

    25 > 26 > 27 > EBB30 > 30 > 31 > 32

c3 forges and briefly selects the chain

    25 > EBB30 > 33              -- NOTE WELL it violates the "k+ blocks in 2k slots" invariant

c4 selects that chain, but is unable to follow c3 when c3 switches to the
superior primary chain.

(We do not think the fact that it forges an EBB is crucial to the Issue. This is unfortunately the only repro we have yet observed; the choice of when exactly to forge an EBB is unspecified, so it'd be nice to empirically confirm that those details are irrelevant. We are only forging EBBs in later epoches in the tests, to more thoroughly exercise various code paths that involve them.)

In more detail:

1. c3 and c4 join at the onset of slot 33. c3 downloads headers and fetches
   blocks from c2 (ie it's only connection to the existing net). Once it
   downloads a header, fetches the block, and selects the block, it can then
   relay the header to c4, who might then fetch the block, and might then
   select it.

2. However, c3 also leads slot 33. Since k = 3, slot 33 is in the third epoch.
   Because of our test infrastructure, c3 will not forge until it observes or
   itself forges an epoch 3 EBB. In this case, c3 will forge and add an EBB 3
   as soon as c3 selects the block from slot 25: this is the soonest that an
   EBB in slot 30 would be within 2k of the tip of the chain. Once it forges
   that EBB, c3 immediately forges its new block for slot 33. So, almost
   immediately after it selects the block in slot 25, its chain has the suffix
   25 :> 30 :> 33.

3. c4 follows c3 along this new suffix; it downloads the headers, fetches and
   ultimately selects the blocks 25 :> 30 :> 33.

4. c3 now realizes its mistake. Shortly after it forges the block in slot 33,
   it's CS and BF clients for the peer c2 fetch the rest of the net's existing
   chain: 25 :> 26 :> 27 :> ... and so on. As soon as it fetches the block in
   slot 27, it abandons its own fork and switches to that one, since it has a
   higher block number than the block in slot 33 (since the block in slot 30 is
   an EBB, and EBBs do not increment the block number).

5. Now that is has switched to the superior chain, it tells c4 to rollback its
   c3 candidate chain to slot 25, and then tells its to roll that candidate
   chain forward to slot 27. This is where the code first realizes something
   has gone wrong. c4 can't roll the candidate chain forward to slot 27,
   because it can't validate that header, because c4's current tip is slot 33
   and 26 + 2k < 33. So c4's CS client raises InvalidRollForward.

6. c4 will never be able to validate the block in slot 26, and once c3 adopts
   it, it's immutable on every node in the database except for c4. So c4 will
   no longer be able to chain sync with the net.

**Comment by mrBliss** [comment]
Fixed in #1942, but #1941 is about removing it again :confused:.
